File: pluginbase-core/messages/src/main/java/pluginbase/messages/DefaultMessage.java
package pluginbase.messages;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.xml.sax.SAXException;

import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.parsers.ParserConfigurationException;
import java.io.IOException;
import java.net.URL;
import java.util.Enumeration;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

class DefaultMessage implements Message {

    static {
        // Load the theme from theme.xml
        final DocumentBuilderFactory dbFactory = DocumentBuilderFactory.newInstance();
        try {
            Enumeration<URL> urls = Messages.class.getClassLoader().getResources(Theme.getThemeResource());
            if (urls.hasMoreElements()) {
                try {
                    DocumentBuilder documentBuilder = dbFactory.newDocumentBuilder();
                    try {
                        Theme.loadTheme(documentBuilder.parse(urls.nextElement().openStream()));
                    } catch (SAXException | IOException e) {
                        e.printStackTrace();
                    }
                } catch (ParserConfigurationException e) {
                    e.printStackTrace();
                }
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    @NotNull
    private final String def;
    @Nullable
    private final String[] key;
    private final int argCount;

    DefaultMessage(@Nullable final String key, @NotNull final String def) {
        if (key != null) {
            this.key = key.split("\\.");
        } else {
            this.key = null;
        }
        this.def = Theme.parseMessage(def);
        this.argCount = countArgs(this.def);
    }

    DefaultMessage(@NotNull final String def) {
        this.key = null;
        this.def = Theme.parseMessage(def);
        this.argCount = countArgs(this.def);
    }

    private static final Pattern PATTERN = Pattern.compile("%s");

    static int countArgs(@NotNull final String def) {
        final Matcher matcher = PATTERN.matcher(def);
        int count = 0;
        while (matcher.find()) {
            count++;
        }
        return count;
    }

    /**
     * The default message in whatever your plugin's primary language is.
     *
     * @return The default non-localized messages.
     */
    @NotNull
    public String getDefault() {
        return def;
    }

    /**
     * The localization key for the message.
     *
     * @return The localization key for the message.
     */
    @Nullable
    public Object[] getKey() {
        return key;
    }

    /**
     * Gets the number of expected arguments for this message.
     * <p/>
     * This is used to validate localized versions of this message to ensure they were given the appropriate
     * amount of arguments.
     *
     * @return the number of expected arguments for this message.
     */
    public int getArgCount() {
        return argCount;
    }

    @Override
    public String toString() {
        return "Message{" +
                "def='" + def + '\'' +
                ", key='" + key + '\'' +
                ", argCount=" + argCount +
                '}';
    }
}

File: pluginbase-core/messages/src/main/kotlin/pluginbase/messages/DefaultMessage.kt
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
package pluginbase.messages

import org.xml.sax.SAXException
import pluginbase.logging.Logging
import java.io.IOException
import javax.xml.parsers.DocumentBuilderFactory
import javax.xml.parsers.ParserConfigurationException

/**
 * A localization key and its defaults.
 *
 * The key represents the location of the localized strings in a language file.
 *
 * The default is what should populate the localization file by default.
 */
internal class DefaultMessage internal constructor (key: String?, default: String) : Message {

    /**
     * The default message in whatever your plugin's primary language is.
     *
     * @return The default non-localized messages.
     */
    override val default: String = Theme.parseMessage(default)

    /**
     * The localization key for the message. Each object in the array represents a configuration node path.
     *
     * @return The localization key for the message.
     */
    override val key: Array<Any>? = key?.split(".")?.dropLastWhile { it.isEmpty() }?.toTypedArray()

    /**
     * Gets the number of expected arguments for this message.
     *
     * This is used to validate localized versions of this message to ensure they were given the appropriate
     * amount of arguments.
     *
     * @return the number of expected arguments for this message.
     */
    override val argCount: Int = Messages.countArgs(this.default)

    /**
     * Bundles this message with preset arguments.
     *
     * Can be used in cases where you are required to return a message of some sort and it is otherwise impossible to
     * return a localized message due to also requiring arguments.
     *
     * Should the number of [args] not match the expected arg count for this [Message], a warning will be logged to
     * help determine where this occurred.
     *
     * @param args The arguments for the bundled message.
     */
    override fun bundle(vararg args: Any?): BundledMessage {
        if (args.size != argCount) {
            Logging.warning("Bundled message created without appropriate number of arguments!")
            for (e in Thread.currentThread().stackTrace) {
                Logging.warning(e.toString())
            }
        }
        return BundledMessage(this, args)
    }

    override fun toString(): String {
        return "Message{def='$default', key='$key', argCount='$argCount'}"
    }

    companion object {

        init {
            // Load the theme from theme.xml
            val dbFactory = DocumentBuilderFactory.newInstance()
            try {
                val urls = Messages::class.java.classLoader.getResources(Theme.themeResource)
                if (urls.hasMoreElements()) {
                    try {
                        val documentBuilder = dbFactory.newDocumentBuilder()
                        try {
                            Theme.loadTheme(documentBuilder.parse(urls.nextElement().openStream()))
                        } catch (e: SAXException) {
                            e.printStackTrace()
                        } catch (e: IOException) {
                            e.printStackTrace()
                        }
                    } catch (e: ParserConfigurationException) {
                        e.printStackTrace()
                    }
                }
            } catch (e: IOException) {
                e.printStackTrace()
            }
        }
    }
}